apollo-router/src/spec/query.rs

```diff
@@ -123,6 +123,19 @@ impl Query {
                                 errors: Vec::new(),
                                 nullified: Vec::new(),
                             };
+                            // Detect if root __typename is asked in the original query (the qp doesn't put root __typename in subselections)
+                            // cf https://github.com/apollographql/router/issues/1677
+                            let operation_kind_if_root_typename =
+                                original_operation.and_then(|op| {
+                                    op.selection_set
+                                        .iter()
+                                        .any(|f| f.is_typename_field())
+                                        .then(|| *op.kind())
+                                });
+                            if let Some(operation_kind) = operation_kind_if_root_typename {
```

Comment:
I think this should check that the defer fragment is at the root, right? if you have  `query { something ...@defer { __typename id} }` would `"__typename": "Query"`  be added to the root of output?

Reply:
Yes Check the tests. I added a similar test